beacon-chain/db/db.go

```diff
@@ -0,0 +1,135 @@
+package db
+
+import (
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/prysmaticlabs/prysm/beacon-chain/types"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/database"
+	"github.com/sirupsen/logrus"
+)
+
+var log = logrus.WithField("prefix", "blockchain")
+
+// BeaconDB manages the data layer of the beacon chain implementation.
+// The exposed methods do not have an opinion of the underlying data engine,
+// but instead reflect the beacon chain logic.
+// For example, instead of defining get, put, remove
+// This defines methods such as getBlock, saveBlocksAndAttestations, etc.
+type BeaconDB struct {
+	db    *database.DB
+	state *beaconState
+}
+
+// Config exposes relevant config options for starting a database.
+type Config struct {
+	Path        string
+	Name        string
+	InMemory    bool
+	GenesisJSON string
+}
+
+type beaconState struct {
```

Comment:
I also think that the state should just be stored in the database for the sake of simplicity, and to optimize once we've perf tested state transitions. Currently, active state and crystallized state are stored in memory and the database, making this area of the code somewhat confusing.

Reply:
We can discuss this separately. For now, I've moved over the original behavior without change.

Reply:
Ah sorry I think I misread your point. yeah crystallized state could be close to 400 mb with a millions validators as mentioned in the last sharding call. Keeping all that in memory may not be the best thing. since crystallized state changes once every 64 slots we can keep that in the db. But the active state should stay in memory

Reply:
One point to keep in mind is that LevelDB/Bolt/Badger all keep recently accessed objects in memory, just not in the application layer. So even if we access active state from the database, it might not result in disk IO.